### DISCUSSION TITLE
Websocket Fragmentation - Bug?

### DISCUSSION THREAD

**Discussion by phkomma** [cesanta/mongoose#1881] [Ask questions, get answers]
Hi,

recently, I switched from mongoose v6 to mongoose v7. For mongoose v6, I implemented websocket packet fragment merging on my own by means of the RFC 6455 standard. In mongoose v7 I got errors within my application when the browser decided to split websocket data into smaller fragments. After delving into the code, I now assume that fragment merging is done by the mongoose library itself. Is this correct? So there is no need of keeping track of fragmented data? If so, why is the 7th bit of websocketMessage->flags, the FIN bit, not set for the case when mongoose merged split data and provides the data in websocketMessage->data? Is this behavior intended or just to be ignored as mongoose merges the split data by its own?

I'd appreciate if you could clarify this issue.

Regards
  Philippe

**Comment by cpq** [comment]
@phkomma yes, mongoose defragments packet itself, you don't need to do anything.
Good question about the FIN flag - let us take a look at it; but if you just ignore the FIN flag, is everything working OK for you?

**Reply by phkomma** [reply]
@cpq Thank you for clarification about the fragmentation. By ignoring the FIN flag, everthing works fine. But then, our code won't be RFC 6455 compatible any more as it expects additional fragments when obtaining websocket data with FIN bit set to zero. As a consequence, our autobahn-base test suite will fail. Hence, it would be fine, if you could adapt the FIN bit behavior when processing fragmented websocket data.

Thanks in advance
  Philippe

**Reply by cpq** [reply]
@phkomma , do you mean just setting the FIN in the `flags` attribute when delivering `MG_EV_WS_MSG`?

**Reply by scaprile** [reply]
I think he means that, as I understand from the RFC he is right
https://www.rfc-editor.org/rfc/rfc6455#section-5.4

**Reply by scaprile** [reply]
Though if we deliver an unfragmented packet Mongoose is behaving OK. I think he does need to see the fragments
